crates/bevy_ecs/src/system/commands/command.rs

```diff
@@ -45,91 +45,57 @@ use crate::{
-    fn with_error_handling(
-        self,
-        error_handler: Option<fn(&mut World, CommandError)>,
-    ) -> impl Command
-    where
-        Self: Sized,
-    {
-        move |world: &mut World| {
-            if let Err(error) = self.apply_internal(world) {
-                // TODO: Pass the error to the global error handler if `error_handler` is `None`.
-                let error_handler = error_handler.unwrap_or(|_, error| panic!("{error}"));
-                error_handler(world, error);
-            }
-        }
-    }
+    fn apply(self, world: &mut World) -> T;
 }
 
-impl<F> Command for F
+impl<F, T> Command<T> for F
 where
-    F: FnOnce(&mut World) + Send + 'static,
+    F: FnOnce(&mut World) -> T + Send + 'static,
 {
-    fn apply(self, world: &mut World) -> Result {
-        self(world);
-        Ok(())
+    fn apply(self, world: &mut World) -> T {
+        self(world)
     }
 }
 
-impl<F> Command<Result> for F
-where
-    F: FnOnce(&mut World) -> Result + Send + 'static,
-{
-    fn apply(self, world: &mut World) -> Result {
-        self(world)
+/// Takes a [`Command`] that returns a Result and uses a given error handler function to convert it into
+/// a [`Command`] that internally handles an error if it occurs and returns `()`.
+pub trait HandleError {
+    /// Takes a [`Command`] that returns a Result and uses a given error handler function to convert it into
+    /// a [`Command`] that internally handles an error if it occurs and returns `()`.
+    fn handle_error_with(self, error_handler: fn(&mut World, Error)) -> impl Command;
+    /// Takes a [`Command`] that returns a Result and uses the default error handler function to convert it into
+    /// a [`Command`] that internally handles an error if it occurs and returns `()`.
+    fn handle_error(self) -> impl Command
+    where
+        Self: Sized,
+    {
+        self.handle_error_with(error_handler::default())
     }
 }
 
-/// Necessary to avoid erasing the type of the `CommandError` in
-/// [`EntityCommand::with_entity`](crate::system::EntityCommand::with_entity).
-impl<F> Command<(Result, CommandError)> for F
-where
-    F: FnOnce(&mut World) -> Result<(), CommandError> + Send + 'static,
-{
-    fn apply(self, world: &mut World) -> Result {
-        self(world)?;
-        Ok(())
+impl<C: Command<Result>> HandleError for C {
+    fn handle_error_with(self, error_handler: fn(&mut World, Error)) -> impl Command {
+        move |world: &mut World| match self.apply(world) {
+            Ok(_) => {}
+            Err(err) => (error_handler)(world, err),
+        }
     }
+}
 
-    fn apply_internal(self, world: &mut World) -> Result<(), CommandError> {
-        self(world)
+/// Takes a [`Command`] that returns a [`Result`] with an error that can be converted into the [`Error`] type
+/// and returns a [`Command`] that internally converts that error to [`Error`] (if it occurs).
+pub fn map_command_err<T, E: Into<Error>>(
```

Comment:
Would it be worth combining this with the impl HandleError above with

```
impl<C: Command<Result<T, E>>, T, E: Into<Error>> HandleError for C { 
   ...
}
```
Since it's essentially what we want in the end?

I see this function is only used once in queue_fallible to create a Command compatible with `HandleError`